librz/core/cmd_api.c

```diff
@@ -1890,14 +1976,14 @@ static void cmd_foreach_cmdname(RzCmd *cmd, RzCmdDesc *cd, RzCmdForeachNameCb cb
 	switch (cd->type) {
 	case RZ_CMD_DESC_TYPE_ARGV:
 		if (rz_cmd_desc_has_handler(cd)) {
-			cb(cmd, cd->name, user);
+			cb(cmd, cd, cd->name, user);
 		}
 		break;
 	case RZ_CMD_DESC_TYPE_ARGV_MODES:
 		for (i = 0; i < RZ_ARRAY_SIZE(argv_modes); i++) {
 			if (cd->d.argv_modes_data.modes & argv_modes[i].mode) {
 				char *name = rz_str_newf("%s%s", cd->name, argv_modes[i].suffix);
-				cb(cmd, name, user);
+				cb(cmd, cd, name, user);
```

Comment:
temporarily swap `cd->name = name` and `cd->summary = summary + summary_suffix`. By doing this here the method `rz_cmd_get_help_json` can be much simpler, in particular with regards to printing the `summary` JSON field.

Reply:
probably, but if i need to do that, then i prefer to add a method to generate a RzCmdDesc which outputs the correct information.

Reply:
What do you mean?

Reply:
i can't do this. it's too messy due `const` being in the structure.

Reply:
Ah, `cd->help` is const :(

Reply:
The as you said, probably having a method to "clone" an existing `cd` is the right way. You can probably just allocate a RzCmdDesc and copy all-fields except for help->summary and cd->name. I think that's the "cleaner" way so the caller doesn't have to deal with summary/names itself.

Reply:
yes, but that is heavier than me doing that check

Reply:
performances don't matter here. It's just a help stuff, it can be "slow" (and by slow i mean no-one will ever notice the slowness). The problem is that if you don't do this here, *every* caller of this function has to be sure to handle the things properly, which is super error prone.

Actually, the foreach function can just provide a `const RzCmdDesc *`, instead of `RzCmdDesc + name` , with the right name and summary already prefilled. This way the callers know what they are iterating without having to rely on *internal details*.